From 7c5a67e957faec98ef816748bd2e7ed6c4f23187 Mon Sep 17 00:00:00 2001
From: Ravi B <ravibabu@ti.com>
Date: Wed, 6 Jul 2011 18:52:58 +0530
Subject: [PATCH 0534/1092] musb: Fix for the test mode feature of musb
 controller

This patch fixes the generation of test packet when
musb controller enters into test mode.

Signed-off-by: Ravi B <ravibabu@ti.com>
Signed-off-by: Vaibhav Hiremath <hvaibhav@ti.com>
---
 drivers/usb/musb/musb_core.c       | 3 ---
 drivers/usb/musb/musb_gadget_ep0.c | 3 +++
 drivers/usb/musb/musb_procfs.c     | 3 +++
 drivers/usb/musb/musb_virthub.c    | 4 ++++
 4 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/drivers/usb/musb/musb_core.c b/drivers/usb/musb/musb_core.c
index e3a81aa..3aeccf4 100644
--- a/drivers/usb/musb/musb_core.c
+++ b/drivers/usb/musb/musb_core.c
@@ -358,12 +358,9 @@ static const u8 musb_test_packet[53] = {
 
 void musb_load_testpacket(struct musb *musb)
 {
-	void __iomem	*regs = musb->endpoints[0].regs;
-
 	musb_ep_select(musb, musb->mregs, 0);
 	musb->ops->write_fifo(musb->control_ep,
 			sizeof(musb_test_packet), musb_test_packet);
-	musb_writew(regs, MUSB_CSR0, MUSB_CSR0_TXPKTRDY);
 }
 
 /*-------------------------------------------------------------------------*/
diff --git a/drivers/usb/musb/musb_gadget_ep0.c b/drivers/usb/musb/musb_gadget_ep0.c
index 22d7ef2..1e6276c 100644
--- a/drivers/usb/musb/musb_gadget_ep0.c
+++ b/drivers/usb/musb/musb_gadget_ep0.c
@@ -761,6 +761,9 @@ irqreturn_t musb_g_ep0_irq(struct musb *musb)
 
 			musb_writeb(mbase, MUSB_TESTMODE,
 					musb->test_mode_nr);
+			if (MUSB_TEST_PACKET == musb->test_mode_nr)
+				musb_writew(musb->endpoints[0].regs,
+					MUSB_CSR0, MUSB_CSR0_TXPKTRDY);
 		}
 		/* FALLTHROUGH */
 
diff --git a/drivers/usb/musb/musb_procfs.c b/drivers/usb/musb/musb_procfs.c
index 1dbf8ab..f282fb9 100644
--- a/drivers/usb/musb/musb_procfs.c
+++ b/drivers/usb/musb/musb_procfs.c
@@ -708,6 +708,9 @@ static int musb_proc_write(struct file *file, const char __user *buffer,
 			musb_load_testpacket(musb);
 			musb_writeb(mbase, MUSB_TESTMODE,
 					MUSB_TEST_PACKET);
+			musb_writew(musb->endpoints[0].regs,
+				MUSB_CSR0, MUSB_CSR0_TXPKTRDY);
+			DBG(2, "musb:testmode sending test packet\n");
 		}
 		break;
 
diff --git a/drivers/usb/musb/musb_virthub.c b/drivers/usb/musb/musb_virthub.c
index 12b90b1..dcdba04 100644
--- a/drivers/usb/musb/musb_virthub.c
+++ b/drivers/usb/musb/musb_virthub.c
@@ -403,6 +403,10 @@ int musb_hub_control(
 				goto error;
 			}
 			musb_writeb(musb->mregs, MUSB_TESTMODE, temp);
+			if (wIndex == 4) {
+				musb_writew(musb->endpoints[0].regs,
+					MUSB_CSR0, MUSB_CSR0_TXPKTRDY);
+			}
 			break;
 		default:
 			goto error;
-- 
1.7.11.2

